test: use pytest fixtures to initialize and populate the backends once per run

anjakefala opened the pull request:
This PR adds a `load_data` to the `TestConf` for all of the current backends in `ibis/backends/{backend}/tests/conftest.py`. PR also removes references to `ci/datamgr.py load` from the `docs/contribute/05_backend_tests.md`.

If the database already exists, the database will be dropped, and recreated.
    
Expected that every backend in the future will have a `load_data` function in its `TestConf`.

This PR uses [filelock](https://py-filelock.readthedocs.io/en/latest/) to ensure that the databases are only instantiated once per test run. It handles the case for if `pytest-xdist` is used to create multiple `pytest` threads. With `pytest-xdist` session fixtures are invoked once per thread.

I attempted to follow the instructions [here](https://github.com/ibis-project/ibis/blob/master/docs/contribute/05_maintainers_guide.md#adding-or-changing-dependencies) to add `filelock` as a dependency, but feedback is welcome if there is something specific I should to ensure it is:
* explicitly marked as a testing-dependency only
* added to the conda.lock
    
 As of this code, `ci/datamgr.py load {backend}` will still work.
    
Closes #2952

pep8speaks commented:
Hello @anjakefala! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`ibis/backends/conftest.py`](https://github.com/ibis-project/ibis/blob/3c8c3f7163dcb8a7a8ff465ae8b0e5271cab38ed/ibis/backends/conftest.py):

> [Line 753:80](https://github.com/ibis-project/ibis/blob/3c8c3f7163dcb8a7a8ff465ae8b0e5271cab38ed/ibis/backends/conftest.py#L753): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

anjakefala commented on ibis/backends/conftest.py:
```diff
@@ -258,10 +443,464 @@ def pytest_runtest_call(item):
-    """Return an instance of BackendTest."""
+def backend(request, data_directory, script_directory):
+    """Return an instance of BackendTest, loaded with data."""
+    # TODO: Will request.param ever be a list of backends?
```
TODO should be removed before merge: add an assert or handle the case for list of backends.

saulpw commented on docs/contribute/04_backend_tests.md:
```diff
@@ -74,22 +74,15 @@ Congrats, you now have a PostgreSQL server running and are ready to run tests!
+The backend needs to be populated with test data.
+The data will be loaded automatically, when the test is run, but it needs to be downloaded first.
 
 1.  Download the data
```
This is now a 1-item list so it can be de-listed.

anjakefala commented on .github/workflows/ibis-backends.yml:
```diff
@@ -96,9 +96,6 @@ jobs:
-
       - name: run non-pyspark tests
         if: ${{ matrix.backend.name != 'pyspark' }}
         run: just ci-check -m ${{ matrix.backend.name }} --numprocesses auto --dist=loadgroup
```
What is `just ci-check`?

cpcloud replied:
If you run `just` (it defaults to running `just --list`) inside the ibis repo you can see descriptions of each command:

```
❯ just
just --list
Available recipes:
    bench *args               # run the benchmark suite
    check *args               # run all non-backend tests; additional arguments are forwarded to pytest
    checklinks *args          # check for invalid links in a locally built version of the docs
    ci-check *args            # run pytest for ci; additional arguments are forwarded to pytest
    clean                     # clean untracked files
    default                   # list justfile recipes
    down                      # stop and remove containers; clean up networks and volumes
    download-data owner="ibis-project" repo="testing-data" rev="master" # download testing data
    fmt                       # format code
    lint                      # lint code
    list-backends             # show all backends
    load-data *backends="all" # load data into running backends; requires a running backend
    test +backends            # run the test suite for one or more backends
    typecheck                 # type check code using mypy
    up *backends              # start backends using docker compose and load data; no arguments starts all backends
```

cpcloud commented:
@anjakefala Thanks for this, it's great work.

The reason there are CI failures is because some of the backends have tests that run against backend data, but aren't "backend" tests in that they do not run across all the backends. The non-backend-backend tests and the proper backend tests share the same data. Without loading that data at some point, the tests will fail to find the things they need to run the test.

In an ideal world we only have proper backend tests for things that ultimately call `execute`, but the backend test suite didn't exist before the need to check that ibis expressions returned the right results when executing 😅.

As for how to proceed, I think if possible we can leave the `python ci/datamgr.py` line in CI and work towards moving the backend-specific backend tests to the larger suite. This will take time, since we'll have to go test by test and remove duplicated tests as well as `notimpl`/`notyet`/`never`-mark the ones that don't work on the other backends. It's kind of a grind 😞.

That said, I'm definitely willing to take on a sizable chunk of that work, which is mostly in the postgres, sqlite and clickhouse backends.

cpcloud commented:
Is it possible to use the code in `ibis/backends/conftest.py` in `ci/datamgr.py` to prevent duplication?

cpcloud commented:
Regarding `git clone`, we're already doing that with `just download-data` so I think we can probably call that done :)

anjakefala commented:
> As for how to proceed, I think if possible we can leave the python ci/datamgr.py line in CI and work towards moving the backend-specific backend tests to the larger suite. This will take time, since we'll have to go test by test and remove duplicated tests as well as notimpl/notyet/never-mark the ones that don't work on the other backends. It's kind of a grind :disappointed:.

> That said, I'm definitely willing to take on a sizable chunk of that work, which is mostly in the postgres, sqlite and clickhouse backends.

Ohhhh, that explains it! I am always pretty happy to help with grindwork. It does not feel like a grind to me.

:+1: if you want me to open an issue for this?

cpcloud commented:
Opening an issue sounds good.

The places to start looking are `ibis/backends/{sqlite,postgres,clickhouse}/tests/test_functions.py`. Much of the duplicated backend testing lies there.

I suspect that after we're done deduplicating that CI times will improve by a decent amount.

cpcloud commented:
I will put up a PR that dedups a test as an example 😄

anjakefala commented on ibis/backends/conftest.py:
```diff
@@ -258,10 +445,556 @@ def pytest_runtest_call(item):
+        os.environ.get(
+            'IBIS_TEST_POSTGRES_USER', os.environ.get('PGUSER', 'postgres')
+        )
+        if not kwargs.get("username", None)
```
These `kwarg` checks were added to support users using non-default values for `ci/datamgr.py`. It does add a lot of boiler plate. 

Do we want to continue supporting those `kwargs`? Boiler plate could also be decreased by me assuming that the `kwargs` for all of the values will be set, if there are `kwargs` being set. Then, I could just do an `if kwargs` check, and set variables appropriately.

kszucs commented on ibis/backends/conftest.py:
```diff
@@ -20,6 +30,91 @@
+    Path
+        Test script directory
+    """
+    return Path(__file__).absolute().parent.parent.parent / "ci"
```
Consider to use [.parents](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.parents)?

kszucs commented:
The `ibis/backends/conftest.py` file is already rather heavy. Would it be possible to reorganize the loaders as staticmethods for the respective `TestConf` classes? Something like 

```
class ClickhouseTestConf(...):
    @staticmethod
    def load_data(data_dir: Path, script_dir: Path, **kwargs):
        ...
```

That could help with the credential deduplication as well as encapsulating backend specific logic in a single file.

anjakefala commented:
@kszucs I agree with this! I am trying to understand why the build is failing before refactoring things. Then, I will refactor with your suggested changes.

```
pytest --durations=25 -ra --junitxml=junit.xml --cov=ibis --cov-report=xml:coverage.xml -m postgres --numprocesses auto --dist=loadgroup
```

Errors. `pytest -s -m postgres` passes. This is why there are failures in the CI, while the build was passing for me locally.

The flavour of errors:

```
>       cursor.execute(statement, parameters)                                                                                                                  
E       sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "pg_class_relname_nsp_index"           
E       DETAIL:  Key (relname, relnamespace)=(test_sequence, 2200) already exists.                                                                             
E                                                                                                                                                              
E       [SQL: CREATE SEQUENCE IF NOT EXISTS test_sequence]                                                                                                     
E       (Background on this error at: https://sqlalche.me/e/14/gkpj)    
```

anjakefala commented:
@cpcloud Yes, exactly. The longer `pytest` fails locally, but not the second one. The longer one is the one being run by the CI tests. 

My current guess from the error messages is that the tests are trying to create the same tables multiple times.

anjakefala commented:
I pushed an experiment that uses [FileLock](https://py-filelock.readthedocs.io/en/latest/index.html) to handle multiple concurrent pytest threads. Ideally, it ensures that the database is re-loaded once per run, with other threads waiting until the loading completes. FileLock supports both Posix and Windows.

If this does look good, things that will need to be done:
- [ ] FileLock will needed to be added to the testing dependencies for poetry and conda. 
- [ ] I should ensure that the teardown works if the `pytest` is run multiple times, and crashes mid-way.

anjakefala commented:
I added this to my PR description:

This PR uses [filelock](https://py-filelock.readthedocs.io/en/latest/) to ensure that the databases are only instantiated once per test session. It handles the case for if `pytest-xdist` is used to create multiple `pytest` threads. I attempted to follow the instructions [here](https://github.com/ibis-project/ibis/blob/master/docs/contribute/05_maintainers_guide.md#adding-or-changing-dependencies) to add `filelock` as a dependency, but feedback is welcome if there is something specific I should to ensure it is:
* explicitly marked as a testing-dependency only
* added to the conda.lock

anjakefala commented:
I tried @kszucs suggestion experimentally, and I did get it to work for both `pytest` and `ci/datamgr.py`. It feels like the approach with the least duplication, so I am going to be implementing it. I will then re-open for review.

kszucs commented on ibis/backends/clickhouse/tests/conftest.py:
```diff
@@ -27,6 +28,72 @@ class TestConf(UnorderedComparator, BackendTest, RoundHalfToEven):
+        """
+        import clickhouse_driver
+
+        host = (
```
`kwargs.get("host", CLICKHOUSE_HOST)` should do the same, or you could explicitly define these as default keyword arguments.

kszucs commented on ibis/backends/conftest.py:
```diff
@@ -259,9 +441,29 @@ def pytest_runtest_call(item):
+    # get the temp directory shared by all workers
+    root_tmp_dir = tmp_path_factory.getbasetemp().parent
+
+    fn = root_tmp_dir / "lockfile"
```
```suggestion
    fn = root_tmp_dir / "lockfile.lock"
```

anjakefala replied:
I think what I have is correct!

There is `lockfile` and `lockfile.lock`. If `lockfile.lock` does not exist and `lockfile` does not, create `lockfile.lock`, load data, and then create `lockfile`.
If `lockfile.lock` exists, wait until it does not exist, and then continue.
If `lockfile.lock` does not exist and `lockfile` does, then continue.

kszucs replied:
Ahh right, so `lockfile` itself indicates whether the backend has been initialized or not. 

Is this `lockfile` shared for all backends? Do we need separate lockfiles perhaps?

anjakefala replied:
> Is this lockfile shared for all backends? Do we need separate lockfiles perhaps?

This is a really good point! I think we do.

kszucs commented on ibis/backends/datafusion/tests/conftest.py:
```diff
@@ -16,6 +16,10 @@ class TestConf(BackendTest, RoundAwayFromZero):
     bool_is_int = True
 
+    @staticmethod
+    def load_data(data_dir, script_dir, **kwargs):
```
I suggest to add this default implementation to the base `BackendTest` class.

kszucs commented on ibis/backends/duckdb/tests/conftest.py:
```diff
@@ -2,13 +2,54 @@
         self.connection = self.connect(data_directory)
 
+    @staticmethod
+    def load_data(data_dir, script_dir, **kwargs):
```
```suggestion
    def load_data(data_dir, script_dir, database="ibis_testing", **kwargs):
```

cpcloud replied:
This is fine for now. I will address in a follow up.

kszucs commented on ibis/backends/duckdb/tests/conftest.py:
```diff
@@ -2,13 +2,54 @@
+            if not kwargs.get("schema", None)
+            else kwargs["schema"]
+        )
+        tables = (
```
If we don't define table schemas anywhere else then baked in `TEST_TABLES` would be enough here as well.

kszucs commented on ibis/backends/mysql/tests/conftest.py:
```diff
@@ -48,6 +50,67 @@ def __init__(self, data_directory: Path) -> None:
+            if not kwargs.get("schema", None)
+            else kwargs["schema"]
+        )
+        tables = (
```
tables = list(TEST_TABLES.keys())

kszucs commented on ibis/backends/mysql/tests/conftest.py:
```diff
@@ -48,6 +50,67 @@ def __init__(self, data_directory: Path) -> None:
+        with engine.connect() as con:
+            for table in tables:
+                con.execute(
+                    f"""\
```
I like to use `textwrap.dedent()`  to preserve the tabulation of the python source code.

kszucs commented on ibis/backends/postgres/tests/conftest.py:
```diff
@@ -45,6 +47,70 @@ class TestConf(BackendTest, RoundHalfToEven):
+            Location of scripts defining schemas
+        """
+        user = (
+            PG_USER if not kwargs.get("username", None) else kwargs["username"]
```
Use default keyword arguments instead?

kszucs reviewed (changes requested):
This is really nice, just a couple of nits.

kszucs commented on ibis/backends/conftest.py:
```diff
@@ -259,9 +441,29 @@ def pytest_runtest_call(item):
+    fn = root_tmp_dir / "lockfile"
+    with FileLock(str(fn) + ".lock"):
+        if not fn.is_file():
+            cls.load_data(data_directory, script_directory)
```
Is the lockfile unique across pytest invocations meaning that we load data each time we execute pytest?

anjakefala replied:
Oh, good point! The lockfile is unique, and so as long as it exists, we will not load data each time we execute.

Would the preferred behaviour be to:
* keep that behaviour
* clean up the lockfile in teardown

cpcloud commented on ibis/backends/clickhouse/tests/conftest.py:
```diff
@@ -27,6 +28,50 @@ class TestConf(UnorderedComparator, BackendTest, RoundHalfToEven):
+        password = kwargs.get("password", CLICKHOUSE_PASS)
+        database = kwargs.get("database", IBIS_TEST_CLICKHOUSE_DB)
+
+        schema = open(script_dir / 'schema' / 'clickhouse.sql')
+        tables = list(TEST_TABLES.keys())
+
+        client = clickhouse_driver.Client(
+            host=host,
+            port=port,
+            user=user,
+            password=password,
+        )
+
+        client.execute(f"DROP DATABASE IF EXISTS {database}")
+        client.execute(f"CREATE DATABASE {database}")
+        client.execute(f"USE {database}")
+
+        for stmt in filter(None, map(str.strip, schema.read().split(";"))):
```
```suggestion
        with open(script_dir / 'schema' / 'clickhouse.sql') as schema:
	          tables = list(TEST_TABLES.keys())
	  
	          client = clickhouse_driver.Client(
	              host=host,
	              port=port,
	              user=user,
	              password=password,
	          )
	  
	          client.execute(f"DROP DATABASE IF EXISTS {database}")
	          client.execute(f"CREATE DATABASE {database}")
	          client.execute(f"USE {database}")
	  
	          for stmt in filter(None, map(str.strip, schema.read().split(";"))):
```

cpcloud reviewed (changes requested):
Few changes around file handles needed, but after that this is good to go!

kszucs reviewed (approved):
Just tested it out locally and everything is working as expected, so it looks good to me.

Thanks @anjakefala!

@cpcloud do you want to have a final look?

cpcloud commented on ibis/backends/conftest.py:
```diff
@@ -39,6 +129,98 @@ def data_directory() -> Path:
+    if url.database is not None:
+        with engine.connect() as conn:
+            conn.execute(f'DROP DATABASE IF EXISTS {database}')
+            conn.execute(f'CREATE DATABASE {database}')
```
Looks like this isn't covered, but we can explore (and hopefully delete!) in a follow up.

cpcloud commented on ibis/backends/mysql/tests/conftest.py:
```diff
@@ -48,6 +50,67 @@ def __init__(self, data_directory: Path) -> None:
             self.__class__.supports_window_operations = True
 
+    @staticmethod
+    def load_data(data_dir: Path, script_dir: Path, **kwargs) -> None:
```
Similarly, I will address later.

cpcloud reviewed (approved):
One of my favorite PRs :)

Excellent work.
